pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/steps/SourceIdentifierScannerTest.scala

```diff
@@ -0,0 +1,190 @@
+
+import scala.util.{Failure, Success}
+
+class SourceIdentifierScannerTest
+    extends FunSpec
+    with IdentifiersGenerators
+    with Matchers
+    with SqlIdentifiersGenerators {
+  case class ClassWithIdentifier(
+    sourceIdentifier: SourceIdentifier,
+    classWithIdentifiers: List[ClassWithIdentifier] = Nil)
+  case class ClassWithIdentifierAndIdentifiedType(
+    sourceIdentifier: SourceIdentifier,
+    identifiedType: String,
+    classWithIdentifiers: List[ClassWithIdentifierAndIdentifiedType] = Nil
+  )
+  case class ClassWithIdentifierAndCanonicalId(
+    sourceIdentifier: SourceIdentifier,
+    canonicalId: String,
+    classWithIdentifiers: List[ClassWithIdentifierAndCanonicalId] = Nil)
+  case class ClassWithIdentifierTypeAndCanonicalId(
+    sourceIdentifier: SourceIdentifier,
+    canonicalId: String,
+    `type`: String,
+    classWithIdentifiers: List[ClassWithIdentifierTypeAndCanonicalId] = Nil
+  )
+  describe("scan") {
+    it("retrieves a sourceIdentifiers at the root of the json") {
+      val sourceIdentifier = createSourceIdentifier
+      val objectWithIdentifier = ClassWithIdentifier(sourceIdentifier)
+      SourceIdentifierScanner
+        .scan(objectWithIdentifier.asJson)
+        .get shouldBe List(sourceIdentifier)
+    }
+
+    it("retrieves a sourceIdentifiers nested in the json") {
+      val sourceIdentifiers = (1 to 4).map(_ => createSourceIdentifier)
+      val objectWithIdentifier = ClassWithIdentifier(
+        sourceIdentifiers.head,
+        List(
+          ClassWithIdentifier(sourceIdentifiers(1)),
+          ClassWithIdentifier(
+            sourceIdentifiers(2),
+            List(ClassWithIdentifier(sourceIdentifiers(3)))))
+      )
+      SourceIdentifierScanner
+        .scan(objectWithIdentifier.asJson)
+        .get should contain theSameElementsAs (sourceIdentifiers)
+    }
+
+    it("throws an exception if it cannot parse a sourceIdentifier") {
+      val json =
+        """
+          |{
+          | "sourceIdentifier": {
+          |   "something": "something"
+          | }
+          |}
+          |""".stripMargin
+
+      SourceIdentifierScanner.scan(parse(json).right.get) shouldBe a[Failure[_]]
+
+    }
+  }
+  describe("update") {
+    it("modifies json to add a single canonicalId in the root") {
+      val sourceIdentifier = createSourceIdentifier
+      val objectWithIdentifier = ClassWithIdentifier(sourceIdentifier)
+      val identifier = Identifier(
+        canonicalId = createCanonicalId,
+        sourceIdentifier = sourceIdentifier)
+      val identified = SourceIdentifierScanner.update(
+        objectWithIdentifier.asJson,
+        Map(sourceIdentifier -> identifier))
+      identified shouldBe a[Success[_]]
+      identified.get
+        .as[ClassWithIdentifierAndCanonicalId]
+        .right
+        .get shouldBe ClassWithIdentifierAndCanonicalId(
+        sourceIdentifier,
+        canonicalId = identifier.CanonicalId)
```

Comment:
As above, I'd find the JSON easier to follow than the case classes:

```suggestion
      val jsonString = """
        |{
        |  "sourceIdentifier": {
        |    "identifierType": "miro-image-number",
        |    "value": "A0001234",
        |    "ontologyType": "Work"
        |  },
        |  "identifiedType": "IdentifiedWork"
        |}
      """.stripMargin
      
      val sourceIdentifier = SourceIdentifier(
        identifierType = "miro-image-number",
        value = "A0001234",
        ontologyType = "Work"
      )
      
      val identifier = Identifier(
        canonicalId = "single-canonical-id-test",
        sourceIdentifier = sourceIdentifier
      )
            
      val identified = SourceIdentifierScanner.update(
        parse(jsonString).get,
        identifiers = Map(sourceIdentifier -> identifier)
      )
      identified shouldBe a[Success[_]]
      
      val updatedJsonString = 
      
      assertJsonStringsEqual(
        identified.get.noSpaces,
        """
        {
          "sourceIdentifier": {
            "identifierType": "miro-image-number",
            "value": "A0001234",
            "ontologyType": "Work"
          },
          "canonicalId": "single-canonical-id-test",
          "type": "IdentifiedWork"
        }
        """
      )
```

(Note: as written, I don't think these case classes have an `identifiedType` field. Should they?)

Reply:
re `identifiedType` - there are separate types/classes for with and without because it needs to work for both situations